File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        n = len(candidates)
        path = []
        ans = []
        candidates.sort()
        def dfs(i, left):
            if left == 0:
                ans.append(path.copy())
                return
            if i == n or candidates[i] > left:
                return
            for j in range(i, n):
                #确保同一层不重复
                if j > i and candidates[j] == candidates[j-1]:
                    continue
                path.append(candidates[j])
                dfs(j+1, left - candidates[j])
                path.pop()
        dfs(0, target)
        return ans

File: Data Structures & Algorithms/generate-parentheses/submission-1.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<Character> path = new ArrayList<>();
        List<String> ans = new ArrayList<>();
        dfs(n, path, ans, 0, 0);
        return ans;

    }

    void dfs(int n, List<Character> path, List<String> ans, int left, int right) {
        if (path.size() == n * 2) {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < n * 2; i++) {
                sb.append(path.get(i));
            }
            ans.add(sb.toString());
            return;
        }
        
        if(left < n) {
            path.add('(');
            dfs(n, path, ans, left + 1, right);
            path.remove(path.size() - 1);
        }

        if(left > right && right < n) {
            path.add(')');
            dfs(n, path, ans, left, right + 1);
            path.remove(path.size() - 1);
        }
    }
}
